velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -549,6 +550,95 @@ std::string toString(const MaterializedRow& row) {
   return oss.str();
 }
 
+// Generate user friendly diff message for materialized rows bucketized by
+// tryBucketMaterializedRows. Values are compared with epsilon.
+std::string generateUserFriendlyDiff(
+    const std::vector<MaterializedRow>& expectedRows,
+    const std::vector<MaterializedRow>& actualRows) {
+  std::vector<MaterializedRow> extraActualRows;
+  std::vector<MaterializedRow> missingActualRows;
+  int32_t actualIndex = 0;
+  int32_t expectedIndex = 0;
+
+  auto lessThanWithEpsilon = [](const MaterializedRow& lhs,
+                                const MaterializedRow& rhs) {
+    const auto minSize = std::min(lhs.size(), rhs.size());
+    for (size_t i = 0; i < minSize; ++i) {
+      if (lhs[i].equalsWithEpsilon(rhs[i])) {
+        continue;
+      }
+      // The 1st non-equal element determines if 'left' is smaller or not.
+      return lhs[i].lessThanWithEpsilon(rhs[i]);
+    }
+    // The shorter vector is smaller.
+    return lhs.size() < rhs.size();
+  };
+  auto equalWithEpsilon = [](const MaterializedRow& lhs,
+                             const MaterializedRow& rhs) {
+    if (lhs.size() != rhs.size()) {
+      return false;
+    }
+    auto size = lhs.size();
+    for (auto i = 0; i < size; ++i) {
+      if (!lhs[i].equalsWithEpsilon(rhs[i])) {
+        return false;
+      }
+    }
+    return true;
+  };
+
+  while (expectedIndex < expectedRows.size() &&
+         actualIndex < actualRows.size()) {
+    bool equal = true;
+    if (equalWithEpsilon(
+            expectedRows[expectedIndex], actualRows[actualIndex])) {
+      ++expectedIndex;
+      ++actualIndex;
+    } else if (lessThanWithEpsilon(
+                   expectedRows[expectedIndex], actualRows[actualIndex])) {
+      missingActualRows.push_back(expectedRows[expectedIndex]);
+      ++expectedIndex;
+    } else {
+      extraActualRows.push_back(actualRows[actualIndex]);
+      ++actualIndex;
+    }
+  }
+  while (actualIndex < actualRows.size()) {
+    extraActualRows.push_back(actualRows[actualIndex]);
+    ++actualIndex;
+  }
+  while (expectedIndex < expectedRows.size()) {
+    missingActualRows.push_back(expectedRows[expectedIndex]);
+    ++expectedIndex;
+  }
+
+  std::ostringstream message;
```

Comment:
This logic seems to be duplicated. Consider extracting it into a helper function and reuse.